### ISSUE TITLE
Trying to access '$app-name' on headerless message hangs rsyslog, crashes after multiple kill signals

### ISSUE THREAD

**Issue by StrongestNumber9** [rsyslog/rsyslog#3135]
### Expected behavior
$app-name is equal to string "". Rsyslog keeps on working and/or prints warning messages about what happened.

### Actual behavior
Hangs while running, ignores kill signals until it crashes. No new connections are accepted after this. 

```
5993.095036471:imrelp.c       : parser.c: msg parser: flags 30, from 'localhost', msg '[2018-10-16 09:59:13] ping pong
'
5993.095043724:imrelp.c       : parser.c: parse using parser list 0x7f2bc639c8e0 (the default list).
5993.095050135:imrelp.c       : parser.c: dropped LF at very end of message (DropTrailingLF is set)
5993.095057431:imrelp.c       : parser.c: Parser 'rsyslog.rfc5424' returned -2160
5993.095063697:imrelp.c       : pmrfc3164.c: Message will now be parsed by the legacy syslog parser (offAfterPRI=0, lenMsg=31.
5993.095069436:imrelp.c       : pmrfc3164.c: msg seems to be headerless, treating it as such
5993.095075270:imrelp.c       : parser.c: Parser 'rsyslog.rfc3164' returned 0
5993.095081255:imrelp.c       : ruleset.c: processBATCH: batch of 1 elements must be processed
5993.095087002:imrelp.c       : ruleset.c: processBATCH: next msg 0: [2018-10-16 09:59:13] ping pong
5993.095093764:imrelp.c       : rainerscript.c:     SET !tag =
5993.095105524:imrelp.c       : rainerscript.c:       var 'app-name'
5993.095137015:imrelp.c       : rainerscript.c:     END SET
5993.095149153:imrelp.c       : rainerscript.c: eval expr 0x7f2bc63c8fb0, type 'V'
6104.907796959:main thread    : rsyslogd.c: DoDie called.
6104.907823942:main thread    : rsyslogd.c: exiting on signal 15
6104.907830378:main thread    : rsyslogd.c: Terminating input threads...
6104.907839553:main thread    : ../threads.c: request term via SIGTTIN for input thread 'imrelp' 0x7f2bbf334700
6104.907846496:main thread    : ../threads.c: thread imrelp: initiating termination, timeout 1000 ms
6104.907923949:imrelp.c       : imrelp.c: imrelp: termination requested via SIGTTIN - telling RELP engine
6105.891999093:main thread    : rsyslogd.c: DoDie called.
6105.908035865:main thread    : ../threads.c: input thread term: timeout expired waiting on thread imrelp termination - canceling
6105.908101014:main thread    : ../threads.c: non-cancel input thread termination FAILED for thread imrelp 0x7f2bbf334700
6106.630999519:main thread    : rsyslogd.c: DoDie called.
6107.173448691:main thread    : rsyslogd.c: DoDie called.
6107.661055504:main thread    : rsyslogd.c: DoDie called.
Aborted (core dumped)

```

### Steps to reproduce the behavior
Minimal rsyslog.conf
```
$AbortOnUncleanConfig on
$LocalHostName test
main_queue(queue.type="Direct")
global(
    workDirectory="/path/to/workdir"
    maxMessageSize="128k"
)

module(load="imrelp")
input(type="imrelp" port="601" ruleset="input")

ruleset(name="input") {
    set $!tag = $app-name;
    action(type="omfile" file="/dummy.log")
}
```

```
# rsyslogd -f rsyslog.conf -i rsyslog.pid -N9
# rsyslogd -f rsyslog.conf -i rsyslog.pid -dn
# tcpflood -p 601 -T relp-plain -M "[2018-10-16 09:59:13] ping pong"
# kill $PID
# kill $PID
<snip>
# kill $PID
# kill $PID
```

### Environment
Newest rsyslog from rhel/centos repository
```
# /usr/sbin/rsyslogd -v
rsyslogd 8.38.0, compiled with:
        PLATFORM:                               x86_64-redhat-linux-gnu
        PLATFORM (lsb_release -d):
        FEATURE_REGEXP:                         Yes
        GSSAPI Kerberos 5 support:              Yes
        FEATURE_DEBUG (debug build, slow code): No
        32bit Atomic operations supported:      Yes
        64bit Atomic operations supported:      Yes
        memory allocator:                       system default
        Runtime Instrumentation (slow code):    No
        uuid support:                           Yes
        systemd support:                        Yes
        Number of Bits in RainerScript integers: 64

See http://www.rsyslog.com for more information.
```

Running on centos container on docker
```
# cat /etc/centos-release
CentOS Linux release 7.5.1804 (Core)
```

```
# uname -a
Linux test-machine 3.10.0-514.el7.x86_64 #1 SMP Tue Nov 22 16:42:41 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by rgerhards** [comment]
Can you pls provide a sample message "without" app-name. I ask because rsyslog heuristic should be able to find one, and so having a sample where this does not work is necessary to troubleshoot this.

**Comment by rgerhards** [comment]
sry, overlookd the tcpflood! so all set :-)

**Comment by rgerhards** [comment]
@jgerhards is trying to craft a testbench test for this. We usually require that step to assign the bug label - only a testbench test ensures we will not have future regressions.

**Comment by rgerhards** [comment]
> Hangs while running,

This seems to have been confirmed via a testbench test, so we can go forward in solving.

I need to check if headerless is a problem for RELP. Because RELP expects a valid syslog message.

> ignores kill signals until it crashes.

I guess you run rsyslogd in debug mode. In debug mode, 5 kills during shutdown make rsyslog abort. This is intentional as a way out of a hang condition. We could reproduce this for the test.

**Comment by StrongestNumber9** [comment]
I used imkafka on v8.32 (patched/compiled by @kortemik), I just used imrelp for proof of concept purposes on official v8.38.

Retested/compared our version to official one and noticed that on our version PoC works fine and imkafka hangs randomly, while on official v8.38 both hangs reliably. I think the core problem is still the same though, just reliably exposed in v8.38.

**Comment by rgerhards** [comment]
I can confirm it happens also with imtcp (which is even simpler in regard to debugging and available on more testbench workers). Hope to nail it today.

**Comment by rgerhards** [comment]
I can confirm a deadlock.

**Comment by rgerhards** [comment]
commit https://github.com/rgerhards/rsyslog/commit/997de6f2140230aa6836a209eedc9c2d71abfc9c looks like a proper fix, PR #3140 has some more commits with cleanup refactor and a new testbench test, but these are not strictly necessary. The initial commit already passed CI testing.

@StrongestNumber9  if you like, you can give it a try yourself. If you do, please report back. In any case, I'll merge #3140 if CI does not discover any hidden problem (in the other commits). I know for sure it fixes a potential deadlock.
